src/dipdup/datasources/tzkt/datasource.py

```diff
@@ -812,15 +823,33 @@ async def _extract_message_data(self, type_: MessageType, message: List[Any]) ->
                         raise RuntimeError('Reorg message received, but neither current nor sync level is known')
 
                     # NOTE: This rollback does not affect us, so we can safely ignore it
-                    if current_level <= level:
+                    if current_level <= message_level:
+                        return
+
+                # NOTE: Drop buffered messages with higher level until available
+                rolled_back_levels = range(current_level, message_level, -1)
+                for rolled_back_level in rolled_back_levels:
+                    if not self._buffer.pop(rolled_back_level, None):
```

Comment:
I'd also add info message in "else" case (when rolled back block is in the buffer), for observability